### PR TITLE
feat: open editor when clicking error on overlay

### PR DESCRIPTION
**Pull request by malcolm-kee**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### Motivation / Use-Case

This PR make webpack-dev-server open editor when user clicks on the error message on overlay.

Part of #3689.

### Additional Info

- New endpoint `"/webpack-dev-server/open-editor"` is added on the dev server which the client would calls using `fetch`.
- Right now I assume `moduleIdentifier` from the `stats.errors` is the file name, but I not sure if that assumption is correct.
- Use [`launch-editor`](https://github.com/yyx990803/launch-editor) to open editor (it is based on `react-dev-utils` of Create React App, and is being used by Vite as well).
- For webpack 4 the logic does not work because there is no `moduleIdentifier`, but I assume that is not a dealbreaker.

### CODE REVIEW

File: lib/launchEditor.js [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+      }
+    });
+
+    _childProcess.on("error", (error) => {
+      printInstructions(logger, fileName, error.message);
+    });
+  }
+}
+
+module.exports = launchEditor;
```
Comment (alexander-akait) [inline lib/launchEditor.js:307]: Maybe there is a package for this? Also can we add unit tests for this?

File: lib/Server.js [hunk 2]
```diff
@@ -2033,6 +2034,25 @@ class Server {
       }
     );
 
+    /** @type {import("express").Application} */
+    (app).get(
```
Comment (alexander-akait) [inline lib/Server.js:2038]: Let's add e2e test for this

**Review by alexander-akait — COMMENTED** [review]
Looks good, we need to test it to avoid problems, also maybe we have a package for launching (just question)?

File: lib/launchEditor.js (hunk 1 above)
Comment (malcolm-kee) [reply lib/launchEditor.js]: I've installed [`launch-editor`](https://github.com/yyx990803/launch-editor) package, which apparently is extracted from `react-dev-utils`. It's also used by Vite as well.

File: lib/Server.js (hunk 2 above)
Comment (malcolm-kee) [reply lib/Server.js]: Yup, it's added!

File: test/helpers/run-browser.js [hunk 3]
```diff
@@ -3,6 +3,16 @@
...
 
+/**
+ * @typedef {Object} RunBrowserResult
+ * @property {import('puppeteer').Page} page
+ * @property {import('puppeteer').Browser} browser
+ */
+
+/**
+ * @param {Parameters<import('puppeteer').Page['emulate']>[0]} config
+ * @returns {Promise<RunBrowserResult>}
```
Comment (malcolm-kee) [inline test/helpers/run-browser.js:14]: Additional typing to improve intellisense while writing tests.

File: lib/Server.js [hunk 4]
```diff
@@ -2033,6 +2033,19 @@ class Server {
       }
     );
 
+    /** @type {import("express").Application} */
+    (app).get("/webpack-dev-server/open-editor", (req, res) => {
+      const fileName = req.query.fileName;
+
+      if (typeof fileName === "string") {
+        // @ts-ignore
+        const launchEditor = require("launch-editor");
```
Comment (malcolm-kee) [inline lib/Server.js:2042]: require lazily to make mock possible.

File: client-src/overlay.js [hunk 5]
```diff
@@ -203,6 +203,16 @@ function show(type, messages, trustedTypesPolicyName) {
       typeElement.innerText = header;
       applyStyle(typeElement, msgTypeStyle);
 
+      if (message.moduleIdentifier) {
+        applyStyle(typeElement, { cursor: "pointer" });
+        typeElement.dataset.canOpen = true;
```
Comment (malcolm-kee) [inline client-src/overlay.js:208]: this is hook for e2e test. If there is other preferred convention, let me know!

**Review by alexander-akait — APPROVED** [review]
Sorry for delay, it looks good, @snitin315 Can you review this too?

**Review by snitin315 — APPROVED** [review]
Looks great. Thanks ⭐

**Comment by alexander-akait** [comment]
@malcolm-kee Thank you, do we will have more PRs before release?

**Comment by malcolm-kee** [comment]
@alexander-akait I plan to have one more enhancement for reporting runtime error, but I foreseen that will probably take time as we probably want to allow configuration (e.g. `react-error-overlay` has logic specific to React).

**Comment by alexander-akait** [comment]
Thanks for answer, no rush, I will wait, thank you

**Comment by malcolm-kee** [comment]
@alexander-akait Right now I'm a bit stuck with how to differentiate between error thrown by webpack vs application code, do you have any suggestion?

Let me know if this should be discussed in another channel!

Meanwhile I'll continue try to understand how it was done in `react-error-overlay`.

**Comment by alexander-akait** [comment]
@malcolm-kee Feel free to send a PR how you see to improve it and we will continue dicussion in the PR, thank you

**Comment by nanianlisao** [comment]
@malcolm-kee This function does not seem to work, how should I verify this function

**Comment by malcolm-kee** [comment]
@nanianlisao can you raise an issue with a reproduction? We have test for this so it's apparently working for us.

**Comment by nanianlisao** [comment]
@malcolm-kee Sorry about that, my bad. I thought we could do without the error-overlay-webpack-plugin plugin, but it looks like we're not there yet. We can only catch errors that are passed through the WebSocket for now.

**Pull request merged**